### PR TITLE
Fixed  #3871 (can reproduction when i click 4 times)

### LINKED ISSUE #3871: Triple clicking selects parts of the block below
#### Do you want to request a _feature_ or report a _bug_?

A bug

#### What's the current behavior?

Triple clicking for example a paragraph that has a heading below it also selects part of the heading. This can be seen by looking at the rich text menu, the heading format is selected. When copying and pasting the paragraph, an empty heading is also pasted below it.

This can be reproduced in the rich text demo.

triple-click

Slate: 0.58.4
Browser: Chrome
OS: Mac

#### What's the expected behavior?

Only the block being triple clicked should be selected and copy/pasting it should not paste the empty heading

### PR DESCRIPTION
**Pull request by sennpang**
**Description**
When i click 4 times, the selection is not expected, like #3871 

https://user-images.githubusercontent.com/16861647/163365630-22f55edf-625c-46ad-b575-c4001edc9288.mp4

fixed

https://user-images.githubusercontent.com/16861647/163365819-7f3e7115-cbcc-4bf2-bb40-ff84e5bb54c3.mp4

**Issue**
Fixes: #4930 #3871 

**Example**
https://user-images.githubusercontent.com/82301/162357857-514f3563-a256-45e2-ab11-21f6c019ab0c.mp4

**Context**
In the case of triple clicking on the Editable component itself, the node returned from toSlateNode is the editor node, and the path returned from findPath is of zero length. 

**Checks**
- [x] The new code matches the existing patterns and styles.
- [x] The tests pass with `yarn test`.
- [x] The linter passes with `yarn lint`. (Fix errors with `yarn fix`.)
- [x] The relevant examples still work. (Run examples with `yarn start`.)
- [x] You've [added a changeset](https://github.com/atlassian/changesets/blob/master/docs/adding-a-changeset.md) if changing functionality. (Add one with `yarn changeset add`.)

### CODE REVIEW

File: packages/slate-react/src/components/editable.tsx [hunk 1]
```diff
@@ -760,11 +760,14 @@ export const Editable = (props: EditableProps) => {
...
-                if (event.detail === TRIPLE_CLICK) {
+                if (event.detail >= TRIPLE_CLICK && path.length) {
                   const start = Editor.start(editor, [path[0]])
                   const end = Editor.end(editor, [path[0]])
                   const range = Editor.range(editor, start, end)
-                  Transforms.select(editor, range)
+                  setTimeout(() => {
+                    Transforms.select(editor, range)
+                    Editor.unhangRange(editor, range)
+                  }, 100)
```
Comment (scottlu) [inline packages/slate-react/src/components/editable.tsx:770]: Given the timeout, two potential issues: 1) the range could be invalid by the time the timer expires (editor content could have changed). Calling select() and unhangRange() with an invalid range could cause unexpected behavior including a crash. 2) the component could have unmounted before the timer expires, and these apis should not be called in that case. Making the code synchronous would remove these problems.

#1: I don't know of a fix for this one. It's not clear how you could know the range is valid when the timer fires.
#2: You can check if the component has been unmounted by setting a ref boolean in a useEffect cleanup callback. Then that ref's value can be checked in this setTimeout callback function to see if the component has unmounted. If it has, then don't make these api calls.

**Review by scottlu — CHANGES_REQUESTED** [review]

File: packages/slate-react/src/components/editable.tsx (hunk 1 above)
Comment (sennpang) [reply packages/slate-react/src/components/editable.tsx:771]: add `if (!ref.current) return` in setTimeout callback

Comment (scottlu) [reply packages/slate-react/src/components/editable.tsx:771]: The node, path, and range may not be right after 100ms if editor.children changed within that time. That could be solved by putting all the logic inside the setTimeout callback like this:

```
if (event.detail >= TRIPLE_CLICK) {
  setTimeout(() => {
    if (ref.current) {
      const node = ReactEditor.toSlateNode(editor, event.target)
      const path = ReactEditor.findPath(editor, node)
      if (path.length) {
        const start = Editor.start(editor, [path[0]])
        const end = Editor.end(editor, [path[0]])
        const range = Editor.range(editor, start, end)
        Transforms.select(editor, range)
        Editor.unhangRange(editor, range)
      }
    }
  }, 100);
  return;
}
```

Comment (sennpang) [reply packages/slate-react/src/components/editable.tsx:775]: putting all the logic inside the setTimeout callback have another problem, the event.target maybe not current target , it 
will set selection to prev target
`onDOMSelectionChange` function in `packages/slate-react/src/components/editable.tsx`
image
when i change to `Transforms.unhangRange(editor, range)`, triple and more click is fine, but i don't know  the potential issues

Comment (sennpang) [reply packages/slate-react/src/components/editable.tsx]: @scottlu i delete the setTimeout function, and use ` state.isTripleClick ` to check 3 or more clicks, on `onDOMSelectionChange` event to decide how to set the selection, what do you think?

**Pull request closed without merging**